### PR TITLE
Balder/trace backend

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe PR

### CODE REVIEW

File: staging_vespalib/src/vespa/vespalib/objects/object2slime.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+//-----------------------------------------------------------------------------
+
+void
+Object2Slime::openStruct(const vespalib::string &name, const vespalib::string &type)
+{
+    _stack.push_back(_cursor);
+
+    if (name.empty()) {
+        _cursor = & _cursor->setObject(type);
```
Comment (havardpe) [inline staging_vespalib/src/vespa/vespalib/objects/object2slime.cpp]: please avoid this

File: staging_vespalib/src/vespa/vespalib/objects/object2slime.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (havardpe) [inline staging_vespalib/src/vespa/vespalib/objects/object2slime.h]: old copyright

File: staging_vespalib/src/vespa/vespalib/objects/object2slime.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#include "objectvisitor.h"
+#include <vector>
+
+namespace vespalib {
+
+namespace slime { class Cursor; }
+
+/**
+ * This is a concrete object visitor that will build up a structured
+ * human-readable string representation of an object.
```
Comment (havardpe) [inline staging_vespalib/src/vespa/vespalib/objects/object2slime.h]: update comments

File: staging_vespalib/src/vespa/vespalib/objects/object2slime.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ **/
+class Object2Slime : public ObjectVisitor
+{
+private:
+    slime::Cursor * _cursor;
+    std::vector<slime::Cursor *> _stack;
+
+public:
+    /**
+     * Create an object dumper with the given indent size; default is
```
Comment (havardpe) [inline staging_vespalib/src/vespa/vespalib/objects/object2slime.h]: comment not relevant

**Review by havardpe — COMMENTED** [review]
please consider my comments

File: staging_vespalib/src/vespa/vespalib/objects/object2slime.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/objects/object2slime.cpp]: Fixed

File: staging_vespalib/src/vespa/vespalib/objects/object2slime.h (hunk 2 above)
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/objects/object2slime.h]: Fixed.

File: staging_vespalib/src/vespa/vespalib/objects/object2slime.h (hunk 4 above)
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/objects/object2slime.h]: Fixed

File: staging_vespalib/src/vespa/vespalib/objects/object2slime.h (hunk 3 above)
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/objects/object2slime.h]: Fixed

File: staging_vespalib/src/vespa/vespalib/objects/object2slime.cpp (hunk 1 above)
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/objects/object2slime.cpp]: Fixed.

File: searchlib/src/vespa/searchlib/engine/trace.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/engine/trace.h]: Fixed

File: searchlib/src/vespa/searchlib/engine/trace.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+Trace::~Trace() = default;
+
+Trace::Cursor &
+Trace::createCursor(vespalib::stringref name) {
+    Cursor & trace = _traces.addObject();
+    trace.setString("tag", name);
+    return trace;
+
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/engine/trace.cpp]: Fixed

File: searchlib/src/vespa/searchlib/engine/trace.cpp [hunk 8]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/engine/trace.cpp]: Fixed

File: container-search/src/main/java/com/yahoo/fs4/FS4Properties.java [hunk 9]
```diff
@@ -10,11 +10,13 @@
...
     static public class Entry {
         public final String key;
-        public final String val;
+        private final byte [] val;
         public Entry(byte[] k, byte[] v) {
             key = Utf8.toString(k);
-            val = Utf8.toString(v);
+            val = v;
         }
+        public final String getValueString() { return Utf8.toString(val); }
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/fs4/FS4Properties.java]: Removed, unused

File: container-search/src/main/java/com/yahoo/fs4/FS4Properties.java [hunk 10]
```diff
@@ -10,11 +10,13 @@
...
+        public final byte [] getValue() { return val; }
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/fs4/FS4Properties.java]: Keep as is as only use is a byte array.

File: container-search/src/main/java/com/yahoo/prelude/fastsearch/VespaBackEndSearcher.java [hunk 11]
```diff
@@ -410,9 +415,23 @@ static void traceQuery(String sourceName, String type, Query query, int offset,
         }
     }
 
+    private void addBackendTrace(Query query, QueryResultPacket resultPacket) {
+        if (resultPacket.propsArray == null) return;
+        for (FS4Properties properties : resultPacket.propsArray) {
+            if ( ! properties.getName().startsWith("trace")) continue;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/prelude/fastsearch/VespaBackEndSearcher.java]: 'trace' it is.

File: searchcore/src/vespa/searchcore/proton/matchengine/matchengine.cpp [hunk 12]
```diff
@@ -140,6 +143,14 @@ MatchEngine::performSearch(search::engine::SearchRequest::Source req,
     }
     ret->request = req.release();
     ret->setDistributionKey(_distributionKey);
+    if (ret->request->getTraceLevel() > 0) {
+        vespalib::asciistream os;
+        os << "trace-" << _distributionKey;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matchengine/matchengine.cpp]: Fixed.

File: searchlib/src/tests/queryeval/blueprint/blueprint_test.cpp [hunk 13]
```diff
@@ -698,6 +700,59 @@ getExpectedBlueprint()
            "}\n";
 }
 
+vespalib::string
+getExpectedSlimeBlueprint() {
+    return "{"
+           "    '(anonymous namespace)::MyOr': {"
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/tests/queryeval/blueprint/blueprint_test.cpp]: Fixed as discussed

File: searchlib/src/vespa/searchlib/queryeval/blueprint.cpp [hunk 14]
```diff
@@ -117,6 +119,15 @@ Blueprint::asString() const
     return dumper.toString();
 }
 
+vespalib::slime::Cursor &
+Blueprint::asSlime(const vespalib::slime::Inserter & inserter) const
+{
+    vespalib::slime::Cursor & cursor = inserter.insertObject();
+    vespalib::Object2Slime dumper(cursor);
+    visit(dumper, "", this);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/queryeval/blueprint.cpp]: Fixed as discussed

**Comment by baldersheim** [comment]
@bratseth Could you have a look at the java parts.

**Review by bratseth — APPROVED** [review]
LGTM!

**Pull request merged**